src/Subject.fs

```diff
@@ -113,3 +113,117 @@ module internal Subjects =
                 member this.OnCompletedAsync() = async { OnCompleted |> mb.Post } }
 
         obv, obs
+
+    type private SafeQueue<'a> () =
+        let agent = 
+            MailboxProcessor<QueueMessage<'a>>.Start 
+            <| fun inbox ->
+                let rec loop (queue: Queue<'a>) =
+                    async {
+                        let! message = inbox.Receive()
+                        match message with
+                        | Enqueue item -> 
+                            queue.Enqueue item
+                            return! loop queue
+                        | Dequeue channel ->
+                            let item = queue.Dequeue()
+                            channel.Reply item
+                            return! loop queue
+                        | IterAsync f ->
+                            for msg in queue do 
+                              do! f msg
+                            return! loop queue
+                        | Count channel-> 
+                           channel.Reply (queue.Count)
+                           return! loop queue
+                    }
+                loop <| new Queue<'a>()
+
+        member _.Count () = agent.PostAndReply Count
+        member _.Enqueue item = agent.Post <| Enqueue item
+        member _.Dequeue () = agent.PostAndReply Dequeue
+        member _.IterAsync f = agent.Post(IterAsync f)
+
+    and private QueueMessage<'a> =
+        | Count of AsyncReplyChannel<int>
+        | Enqueue of 'a
+        | Dequeue of AsyncReplyChannel<'a>
+        | IterAsync of ('a -> Async<unit>)
+
+    type private BufferedMessage<'Value> = 
+        | Next of 'Value
+        | Complete
+        | Error of exn
```

Comment:
Could you reuse the existing `Notification<>` type here?

Reply:
Oh duh, I should have. Updated.